File: java/part1/Lines.java
public class Lines 
{
  
  public static void main(String[] args)
  {
    if (args.length!=4)
    {
      System.out.println("Usage: java Lines W H Lmin Lmax");
    }
    else
    {
      
      long w = Integer.parseInt(args[0]); 
      long h = Integer.parseInt(args[1]); 
      long lmin = Integer.parseInt(args[2]); 
      long lmax = Integer.parseInt(args[3]); 
      Lines ls = new Lines(w,h,lmin,lmax);
      ls.lines();
    }
  }
  private long w; 
  private long h; 
  private long lmin; 
  private long lmax; 
  public Lines(long w,long h,long lmin,long lmax)
  {
    this.w = w;
    this.h = h;
    this.lmin = lmin;
    this.lmax = lmax;
  }
  private long box(long w,long h,long dx,long dy)
  {
    return 2*(w+1-dx)*(h+1-dy);
  }
  private void lines()
  {
      long i = 0;
      long j = 0;
      long lminsq=lmin*lmin;
      long lmaxsq=lmax*lmax;
      long acc = 0;
      if (lmin<=1)
      {
        acc=4*w*h+w+h;
      }

      for (i=1,j=2;i<=h;i++)
      {
        for (;j<=w;j++)
        {
          if (accept(lminsq,lmaxsq,j,i))
          {
            acc+=box(w,h,j,i);
          }
        }
        j=1;
      }
      System.out.println(acc);
  }
  private boolean accept(long lminsq,long lmaxsq,long dx,long dy)
  {
    long length = dx*dx+dy*dy;
    return  (((length >= lminsq) && (length <= lmaxsq)) && (gcd(dx,dy)==1));
  }
  private long gcd(long x,long y)
  {
    if (y==0) 
      return x;
    else
      return gcd(y,x%y);
  }

}

File: ml/part1/lines.ml
fun gcd (0,y) = y
  | gcd (x,y) = gcd (y mod x, x)
fun init x y =4*(x)*(y)+x+y

fun box x y dx dy =  Int.toLarge(2*(x+1-dx)*(y+1-dy))


fun accept lminsq lmaxsq dx dy = 
  let 
    val mhkos = dx*dx+dy*dy
  in
    (lminsq <= mhkos) andalso (lmaxsq >= mhkos) andalso (gcd(dx,dy)=1)
  end


fun lines w h lmin lmax = 
  let
    val start = if lmin <= 1 then init w h else 0
    val lminsq=lmin*lmin
    val lmaxsq=lmax*lmax
    fun linesH w h dx dy lmin lmax (acc:IntInf.int) = 
      if w<dx andalso h<dy 
        then acc
        else
          let
            val (nextdx,nextdy)=if dx<=w then ((dx+1),dy) else (1,dy+1)
          in
            if accept lminsq lmaxsq dx dy 
              then linesH w h nextdx nextdy lmin lmax (acc+(box w h dx dy))
              else linesH w h nextdx nextdy lmin lmax acc
          end
  in
    linesH w h 2 1 lmin lmax (Int.toLarge(start))
  end
    


(* MLTON STUFF *)


fun main () = 
  let
    val b = (CommandLine.arguments())
    val st = if (length(b) = 4 )
    then
      let
        val w = Int.fromString (hd(b))
        val h = Int.fromString (hd(tl(b)))
        val lmin = Int.fromString (hd(tl(tl(b))))
        val lmax = Int.fromString (hd(tl(tl(tl(b)))))
        val ln = lines (Option.valOf(w)) (Option.valOf(h)) (Option.valOf(lmin)) (Option.valOf(lmax))
      in
        LargeInt.toString ln
      end
    else
        "Usage: lines W H Lmin Lmax"
  in
    print (st^"\n")
  end
val _ = main ()
